silence/silence_test.go

```diff
@@ -197,19 +197,34 @@ func TestSilencesSetSilence(t *testing.T) {
 		},
 	}
 
-	var called bool
+	done := make(chan bool)
 	s.gossip = &mockGossip{
 		broadcast: func(d mesh.GossipData) {
 			data, ok := d.(*gossipData)
-			require.True(t, ok, "gossip data of unknown type")
-			require.Equal(t, want, data, "unexpected gossip broadcast data")
 
-			called = true
+			// Double check that we can take a lock on s.mtx here.
+			s.mtx.Lock()
+			defer s.mtx.Unlock()
+
+			require.True(t, ok, "gossip data of unknown type")
+			require.Equal(t, want.data, data.data, "unexpected gossip broadcast data")
+			close(done)
 		},
 	}
+
+	// setSilence() is always called with s.mtx locked()
+	s.mtx.Lock()
 	require.NoError(t, s.setSilence(sil))
-	require.True(t, called, "GossipBroadcast was not called")
+	s.mtx.Unlock()
```

Comment:
While this test passes because of the `go`, if someone were to remove the `go`, this deadlocks and the timeout in the `select` below is never called. If you re-arrange the code like below, then we have a passing test with your code :raised_hands: , and if someone removes the `go`, it will fail instead of locking up:

```patch
diff --git a/silence/silence_test.go b/silence/silence_test.go
index 420a78c..eed4060 100644
--- a/silence/silence_test.go
+++ b/silence/silence_test.go
@@ -213,11 +213,11 @@ func TestSilencesSetSilence(t *testing.T) {
 	}
 
 	// setSilence() is always called with s.mtx locked()
-	s.mtx.Lock()
-	require.NoError(t, s.setSilence(sil))
-	s.mtx.Unlock()
-
-	require.Equal(t, want.data, s.st.data, "Unexpected silence state")
+	go func() {
+		s.mtx.Lock()
+		require.NoError(t, s.setSilence(sil))
+		s.mtx.Unlock()
+	}()
 
 	// GossipBroadcast is called in a goroutine.
 	select {
@@ -225,6 +225,8 @@ func TestSilencesSetSilence(t *testing.T) {
 	case <-time.After(1 * time.Second):
 		t.Fatal("GossipBroadcast was not called")
 	}
+
+	require.Equal(t, want.data, s.st.data, "Unexpected silence state")
 }
 
 func TestSilenceSet(t *testing.T) {
```

Reply:
good point !